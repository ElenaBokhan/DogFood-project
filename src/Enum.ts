export enum ETestId {
    ADD_PRODUCT_NAME_INPUT = 'addProductNameInput',
    ADD_PRODUCT_PRICE_INPUT = 'addProductPriceInput',
    ADD_PRODUCT_DESCRIPTION_INPUT = 'addProductDescriptionInput',
    ADD_PRODUCT_DISCOUNT_INPUT = 'addProductDiscountInput',
    ADD_PRODUCT_STOCK_INPUT = 'addProductStockInput',
    ADD_PRODUCT_WIGHT_INPUT = 'addProductWightInput',
    ADD_PRODUCT_PICTURES_INPUT = 'addProductPicturesInput',
    ADD_PRODUCT_SUBMIT_BUTTON = 'addProductSubmitButton',

    ADD_REVIEW_SUBMIT_BUTTON = 'addReviewSubmitButton',
    ADD_REVIEW_TEXT_INPUT = 'addReviewTextInput',

    ADD_TO_FAVOURITES = 'addToFavouritesButton',
    REMOVE_FROM_FAVOURITES = 'removeFromFavourites',

    BREADCRUMBS_BUTTON = 'breadcrumbsButton',

    FOOTER_MAIN_LOGO = 'footerMainLogo',
    FOOTER_MENU = 'footerMenu',
    FOOTER_CONTACTS = 'footerContacts',

    HEADER_MAIN_LOGO = 'headerMainLogo',
    HEADER_ADD_NEW_ICON = 'headerAddNewProduct',
    HEADER_FAVOURITES_ICON = 'headerFavouritesIcon',
    HEADER_CART_ICON = 'headerCartIcon',
    HEADER_PROFILE_ICON = 'headerProfileIcon',
    HEADER_LOG_OUT_ICON = 'headerLogOutIcon',
    HEADER_SEARCH_INPUT = 'headerSearchInput',

    SIGN_FORM_EMAIL_INPUT = 'signFormEmailInput',
    SIGN_FORM_PASSWORD_INPUT = 'signFormPasswordlInput',
    SIGN_FORM_SUBMIT_BUTTON = 'signFormSubmitButton',

    TITLE_PAGE = 'titlePage',
    NOT_FOUND = 'notFound',

    PRODUCT_LIKE_BUTTON = 'likeButton',
    PRODUCT_TRASH_BUTTON = 'trashButton',
    PRODUCT_NAME = 'productName',
    PRODUCT_IMAGE = 'productImage',
}
