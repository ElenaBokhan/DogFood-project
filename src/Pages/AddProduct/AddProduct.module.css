.add-review-page {
padding: 0 120px;
}

.review-textarea {
    height: 88px;
    padding: 14px 16px;
    border-radius: 12px;
    border: 1px solid #CFD8DC;
}

.field {
    display: flex;
    gap: 40px;
    margin: 40px 0;
}

.field span {
    width: 30%;
}

.field input {
    width: 70%;
}

