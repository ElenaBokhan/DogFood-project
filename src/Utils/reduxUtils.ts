import type {AnyAction, AsyncThunk} from '@reduxjs/toolkit';

type GenericAsyncThunk = AsyncThunk<unknown, unknown, any>;
type PendingAction = ReturnType<GenericAsyncThunk['pending']>;
type RejectedAction = ReturnType<GenericAsyncThunk['rejected']>;
type FulfilledAction = ReturnType<GenericAsyncThunk['fulfilled']>;

const hasPrefix = (action: AnyAction, prefix: string): boolean => action.type.startsWith(prefix);
const isPending = (action: PendingAction): boolean => action.type.endsWith('/pending');
const isFulfilled = (action: FulfilledAction): boolean => action.type.endsWith('/fulfilled');
const isRejected = (action: RejectedAction): boolean => action.type.endsWith('/rejected');

export const isActionPending = (prefix: string) => (action: PendingAction) => {
    return hasPrefix(action, prefix) && isPending(action);
};

export const isActionRejected = (prefix: string) => (action: RejectedAction) => {
    return hasPrefix(action, prefix) && isRejected(action);
};

export const isActionFulfilled = (prefix: string) => (action: FulfilledAction) => {
    return hasPrefix(action, prefix) && isFulfilled(action);
};
