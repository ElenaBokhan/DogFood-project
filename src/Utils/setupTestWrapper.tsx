import {router} from 'App';
import {render} from '@testing-library/react';
import {Provider} from 'react-redux';
import {createMemoryRouter, RouterProvider} from 'react-router-dom';
import {ToastContainer} from 'react-toastify';
import store from 'Store/configureStore';

interface WrapperForTestsParams {
    initialRouterEntries: string[];
}

export const setupTestsWrapper = ({initialRouterEntries}: WrapperForTestsParams) => {
    const memoryRouter = createMemoryRouter(router, {
        initialEntries: initialRouterEntries,
    });
    render(
        <Provider store={store}>
            <RouterProvider router={memoryRouter} />
            <ToastContainer autoClose={2000} position="top-right" theme="light" closeOnClick />
        </Provider>
    );

    return memoryRouter;
};
