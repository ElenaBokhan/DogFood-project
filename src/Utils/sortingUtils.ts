export const calculateOldPrice = (price: number, discount: number) => {
    return Math.floor(price - (price * discount) / 100);
};
