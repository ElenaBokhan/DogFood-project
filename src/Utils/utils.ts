import {PER_PAGE} from 'Const';
import {ESortFilter} from 'Pages/Catalog/Catalog';

export const calculateOldPrice = (price: number, discount: number) => {
    return Math.floor(price - (price * discount) / 100);
};

export const isFavourite = (likes: string[], id: string) => {
    return likes.includes(id);
};

export const getDefaultFilter = (): IClientFilter => {
    return {search: '', page: 1, perPage: PER_PAGE};
};

export const dateFormatter = new Intl.DateTimeFormat('ru-RU', {
    year: 'numeric',
    month: 'short',
    day: '2-digit',
    // hour: '2-digit',
    // minute: '2-digit',
});

export const objectHasProperty = <P extends PropertyKey>(obj: unknown, prop: P): obj is object & Record<P, unknown> => {
    return typeof obj === 'object' && !!obj && Object.hasOwn(obj, prop);
};

export const getResponse = (ms: number) => {
    return new Promise((resolve, reject) => {
        setTimeout(() => {
            if (Math.random() > 0.5) {
                resolve('success');
            } else {
                reject(new Error('error'));
            }
        }, ms);
    });
};

export const getSortingProducts = (productList: IProduct[], sortFilter: ESortFilter): IProduct[] => {
    const sortProducts: IProduct[] = [...productList];

    switch (sortFilter) {
        case ESortFilter.POPULAR:
            break;
        case ESortFilter.NEW:
            break;
        case ESortFilter.CHEAPER_FIRST:
            return sortProducts.sort((first, second) => first.price - second.price);
        case ESortFilter.EXPENSIVE_FIRST:
            return sortProducts.sort((first, second) => second.price - first.price);
        case ESortFilter.DISCOUNT:
            return sortProducts.sort((first, second) => second.discount - first.discount);
        case ESortFilter.RATING:
            break;
    }

    return sortProducts;
};
