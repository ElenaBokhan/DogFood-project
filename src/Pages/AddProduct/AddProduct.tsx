import {Button} from 'Components/Common/Button/Button';
import {TitlePage} from 'Components/Common/TitlePage/TitlePage';
import styles from 'Pages/AddReview/AddReview.module.css';
import {useEffect, useState} from 'react';
import {useAddProductMutation} from 'Store/Api/productListApi';
import {Gap} from 'Components/Common/Gap/Gap';
import {FieldForm} from 'Components/Common/FieldForm/FieldForm';
import {useNavigate} from 'react-router-dom';
import {ETestId} from 'Enum';

export interface IProductMap {
    label: string;
    value: string | number;
    param: keyof INewProduct;
    testId: ETestId;
}

export const AddProduct = () => {
    const [addProduct, {isSuccess, data}] = useAddProductMutation();
    const navigate = useNavigate();

    const getInitialProduct = (): IProductMap[] => [
        {
            param: 'name',
            label: 'Название товара',
            value: '',
            testId: ETestId.ADD_PRODUCT_NAME_INPUT,
        },
        {
            param: 'price',
            label: 'Цена товара',
            value: '',
            testId: ETestId.ADD_PRODUCT_PRICE_INPUT,
        },
        {
            param: 'description',
            label: 'Описание',
            value: '',
            testId: ETestId.ADD_PRODUCT_DESCRIPTION_INPUT,
        },
        {
            param: 'discount',
            label: 'Скидка',
            value: '',
            testId: ETestId.ADD_PRODUCT_DISCOUNT_INPUT,
        },
        {
            param: 'stock',
            label: 'Наличие',
            value: '',
            testId: ETestId.ADD_PRODUCT_STOCK_INPUT,
        },
        {
            param: 'wight',
            label: 'Вес',
            value: '',
            testId: ETestId.ADD_PRODUCT_WIGHT_INPUT,
        },
        {
            param: 'pictures',
            label: 'Добавьте фото',
            value: '',
            testId: ETestId.ADD_PRODUCT_PICTURES_INPUT,
        },
    ];

    const [product, setProduct] = useState<IProductMap[]>(getInitialProduct());

    useEffect(() => {
        isSuccess && navigate(`/product/${data._id}`);
    }, [data?._id, isSuccess, navigate]);

    const isNotEmptyValues = (): boolean => {
        for (const field of product) {
            if (field.value === '') return false;
        }

        return true;
    };

    const getNewProduct = (): INewProduct => {
        const newProduct: any = {}; // TODO: разобраться с типизацией

        product.forEach(({value, param}) => {
            newProduct[param] = value;
        });

        return newProduct;
    };

    const handleSubmit = (event: React.SyntheticEvent<HTMLFormElement>) => {
        if (isNotEmptyValues()) {
            addProduct(getNewProduct());
        }
        event.preventDefault();
    };

    const handleSetProduct = (value: string | number, param: keyof INewProduct) => {
        const currentValue = param === 'price' || param === 'discount' || param === 'stock' ? Number(value) : value;

        const newProducts: IProductMap[] = product.map((item) =>
            item.param === param ? {...item, value: currentValue} : item
        );
        setProduct(newProducts);
    };

    const renderAddProductForm = () => {
        return (
            <form onSubmit={handleSubmit}>
                {product.map((field) => (
                    <FieldForm key={field.label} onChange={handleSetProduct} field={field} />
                ))}

                <Button testId={ETestId.ADD_PRODUCT_SUBMIT_BUTTON} type="submit" label={'Добавить продукт'} />
            </form>
        );
    };

    return (
        <div className={styles.addReviewPage}>
            <TitlePage label={'Добавление нового товара'} />
            <Gap size={40} />
            <hr />
            {renderAddProductForm()}
        </div>
    );
};
