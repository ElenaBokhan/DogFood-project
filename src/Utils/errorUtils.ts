import {objectHasProperty} from 'Utils/utils';

export const getMessageFromError = (error: unknown, defaultErrorMessage: string) => {
    if (
        objectHasProperty(error, 'data') &&
        objectHasProperty(error.data, 'message') &&
        typeof error.data.message === 'string'
    )
        return error.data.message;
    return defaultErrorMessage;
};
