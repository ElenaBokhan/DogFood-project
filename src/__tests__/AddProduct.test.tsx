import '@testing-library/jest-dom';
import {setupTestsWrapper} from 'Utils/setupTestWrapper';
import {fireEvent, render, screen} from '@testing-library/react';
import {Provider} from 'react-redux';
import store from 'Store/configureStore';
import {BrowserRouter, createMemoryRouter} from 'react-router-dom';
import {ETestId} from 'Enum';
import {userEvent} from '@testing-library/user-event';
import {AddProduct} from 'Pages/AddProduct/AddProduct';

describe('Тесты для компонента addProduct', () => {
    let router: ReturnType<typeof createMemoryRouter> | undefined;
    let submitBtn: HTMLButtonElement | undefined;
    let nameInput: HTMLInputElement | undefined;
    let priceInput: HTMLInputElement | undefined;
    let descriptionInput: HTMLInputElement | undefined;
    let discountInput: HTMLInputElement | undefined;
    let stockInput: HTMLInputElement | undefined;
    let wightInput: HTMLInputElement | undefined;
    let picturesInput: HTMLInputElement | undefined;
    let allInputs: HTMLInputElement[] | undefined;

    beforeEach(() => {
        router = setupTestsWrapper({initialRouterEntries: ['/addProduct']});
        submitBtn = screen.getByTestId(ETestId.ADD_PRODUCT_SUBMIT_BUTTON);
        nameInput = screen.getByTestId(ETestId.ADD_PRODUCT_NAME_INPUT);
        priceInput = screen.getByTestId(ETestId.ADD_PRODUCT_PRICE_INPUT);
        descriptionInput = screen.getByTestId(ETestId.ADD_PRODUCT_DESCRIPTION_INPUT);
        discountInput = screen.getByTestId(ETestId.ADD_PRODUCT_DISCOUNT_INPUT);
        stockInput = screen.getByTestId(ETestId.ADD_PRODUCT_STOCK_INPUT);
        wightInput = screen.getByTestId(ETestId.ADD_PRODUCT_WIGHT_INPUT);
        picturesInput = screen.getByTestId(ETestId.ADD_PRODUCT_PICTURES_INPUT);
        allInputs = [nameInput, priceInput, descriptionInput, discountInput, stockInput, wightInput, picturesInput];
    });

    describe('Проверяем доступность элементов компонента AddProduct', () => {
        test('Проверяем что все поля доступны', () => {
            checkInputsEnabled(allInputs);
        });

        test('Проверяем что кнопка добавления продукта доступна', () => {
            expect(submitBtn).toBeEnabled();
        });
    });

    describe('Проверяем работу компонента AddProduct при заполненных и незаполлненных полях', () => {
        test('Проверяем что запрос не выполняется при незаполненных полях и редиректа не происходит', async () => {
            checkInputsHasEmptyValue(allInputs);

            await userEvent.click(submitBtn);
            expect(router?.state.location.pathname).toEqual('/addProduct');
        });

        test('Проверяем что поля заполняются верно', async () => {
            const stringInputValue = 'Значение';
            const numberInputValue = 123;

            const stringInputs = [nameInput, descriptionInput, picturesInput, wightInput];
            const numberInputs = [priceInput, stockInput, discountInput];

            fillInputsWithValue(stringInputs, stringInputValue);
            fillInputsWithValue(numberInputs, numberInputValue);

            checkFillInputsValue(stringInputs, stringInputValue);
            checkFillInputsValue(numberInputs, numberInputValue);
        });
    });

    describe('Снапшот компонента AddProduct', () => {
        test('Snapshot', () => {
            const {asFragment} = render(
                <Provider store={store}>
                    <BrowserRouter>
                        <AddProduct />
                    </BrowserRouter>
                </Provider>
            );

            expect(asFragment()).toMatchSnapshot();
        });
    });
});

function checkInputsEnabled(allInputs: HTMLInputElement[]) {
    allInputs.forEach((input) => {
        expect(input).toBeEnabled();
    });
}

function checkInputsHasEmptyValue(allInputs: HTMLInputElement[]) {
    allInputs.forEach((input) => {
        expect(input).toHaveDisplayValue('');
    });
}

async function fillInputsWithValue(allInputs: HTMLInputElement[], value: string | number) {
    allInputs.forEach(async (input) => {
        await fireEvent.change(input, {target: {value}});
    });
}

function checkFillInputsValue(allInputs: HTMLInputElement[], value: string | number) {
    allInputs.forEach(async (input) => {
        expect(input).toHaveDisplayValue(value.toString());
    });
}
