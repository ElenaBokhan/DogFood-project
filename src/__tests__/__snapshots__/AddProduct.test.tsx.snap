// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Тесты для компонента addProduct Снапшот компонента AddProduct Snapshot 1`] = `
<DocumentFragment>
  <div
    class="addReviewPage"
  >
    <div>
      <a
        class="breadcrumbs"
        href="/"
      >
        <img
          alt="arrow"
          src="[object Object]"
        />
        <span
          class="grey undefined subscribe2"
        >
          Назад
        </span>
      </a>
      <span
        class="black undefined head1"
      >
        Добавление нового товара
      </span>
    </div>
    <div
      style="height: 40px;"
    />
    <hr />
    <form>
      <div
        class="field"
      >
        <span
          class="black undefined paragraph1"
        >
          Название товара
        </span>
        <input
          class="fieldTextarea"
          data-testid="addProductNameInput"
          placeholder="Укажите значение"
          type="text"
          value=""
        />
      </div>
      <div
        class="field"
      >
        <span
          class="black undefined paragraph1"
        >
          Цена товара
        </span>
        <input
          class="fieldTextarea"
          data-testid="addProductPriceInput"
          placeholder="Укажите значение"
          type="text"
          value=""
        />
      </div>
      <div
        class="field"
      >
        <span
          class="black undefined paragraph1"
        >
          Описание
        </span>
        <input
          class="fieldTextarea"
          data-testid="addProductDescriptionInput"
          placeholder="Укажите значение"
          type="text"
          value=""
        />
      </div>
      <div
        class="field"
      >
        <span
          class="black undefined paragraph1"
        >
          Скидка
        </span>
        <input
          class="fieldTextarea"
          data-testid="addProductDiscountInput"
          placeholder="Укажите значение"
          type="text"
          value=""
        />
      </div>
      <div
        class="field"
      >
        <span
          class="black undefined paragraph1"
        >
          Наличие
        </span>
        <input
          class="fieldTextarea"
          data-testid="addProductStockInput"
          placeholder="Укажите значение"
          type="text"
          value=""
        />
      </div>
      <div
        class="field"
      >
        <span
          class="black undefined paragraph1"
        >
          Вес
        </span>
        <input
          class="fieldTextarea"
          data-testid="addProductWightInput"
          placeholder="Укажите значение"
          type="text"
          value=""
        />
      </div>
      <div
        class="field"
      >
        <span
          class="black undefined paragraph1"
        >
          Добавьте фото
        </span>
        <input
          class="fieldTextarea"
          data-testid="addProductPicturesInput"
          placeholder="Укажите значение"
          type="text"
          value=""
        />
      </div>
      <button
        class="button standard undefined"
        data-testid="addProductSubmitButton"
        data-value="Добавить продукт"
        type="submit"
      >
        Добавить продукт
      </button>
    </form>
  </div>
</DocumentFragment>
`;
